orttraining/orttraining/python/training/onnxblock/model.py

```diff
@@ -0,0 +1,123 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# model.py
+
+from abc import ABC, abstractmethod
+import onnx
+import onnxruntime.training.onnxblock.model_accessor as accessor
+import onnxruntime.training.onnxblock._graph_utils as graph_utils
+
+
+class Model(ABC):
+    """Builds the forward model based on user's build method."""
+
+    def __init__(self):
+        ...
+
+    @abstractmethod
+    def build(self, *args, **kwargs):
+        """Customize and build the forward graph for this model.
+
+        This method is to be overriden by the user's implementation.
+        """
+        ...
+
+    def __call__(self, *args, **kwargs):
+        """Calls the user's build method and runs validation on top.
+
+        The output onnx model is got by invoking the user's build method.
+        """
+        # build the user model
+        output = self.build(*args, **kwargs)
+
+        # validate and check the model
+        onnx.checker.check_model(accessor.global_accessor.model, True)
+
+        return output
+
+
+class TrainingModel(Model):
+    """Builds the training model based on user's build method."""
+
+    def __init__(self):
+        super(TrainingModel, self).__init__()
+        self._arg_requiring_grad = set()
+        self._arg_not_requiring_grad = set()
+        self._parameters = None
+
+    @abstractmethod
+    def build(self, *args, **kwargs):
+        """Customize and build the forward graph for this training model.
+
+        This method is to be overriden by the user's implementation.
+        """
+        ...
+
+    def requires_grad(self, argument_name, value=True):
+        """Control whether the given graph input/parameter requires gradient."""
+        if value is True:
+            if argument_name in self._arg_not_requiring_grad:
+                self._arg_not_requiring_grad.remove(argument_name)
+            self._arg_requiring_grad.add(argument_name)
+        else:
+            if argument_name in self._arg_requiring_grad:
+                self._arg_requiring_grad.remove(argument_name)
+            self._arg_not_requiring_grad.add(argument_name)
```

Comment:
I assumed, by default, all param should be in self._arg_not_requiring_grad, unless user call requires_grad().

Reply:
is this more clear?

Reply:
My feeling is that the user's model might be very large. And they might need to call `requires_grad` on each and every parameter for the onnx model.
If this process can be automated based on our understanding of the exported model (to decipher which initializer is a model parameter and which is not), we can by default assume all the parameters are trainable. If the user has some parameters that are not trainable, they can set it using `requires_grad(parameter_name, False)`. This is consistent with how one would assume a training model would work (in my opinion). What do you think?